x-pack/metricbeat/module/googlecloud/stackdriver/metricset.go

```diff
@@ -0,0 +1,150 @@
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package stackdriver
+
+import (
+	"context"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"google.golang.org/api/option"
+	monitoringpb "google.golang.org/genproto/googleapis/monitoring/v3"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/cfgwarn"
+	"github.com/elastic/beats/metricbeat/mb"
+	"github.com/elastic/beats/x-pack/metricbeat/module/googlecloud"
+)
+
+const (
+	// MetricsetName is the name of this Metricset
+	MetricsetName = "stackdriver"
+)
+
+// init registers the MetricSet with the central registry as soon as the program
+// starts. The New function will be called later to instantiate an instance of
+// the MetricSet for each host defined in the module's configuration. After the
+// MetricSet has been created then Fetch will begin to be called periodically.
+func init() {
+	mb.Registry.MustAddMetricSet(googlecloud.ModuleName, MetricsetName, New)
+}
+
+// MetricSet holds any configuration or state information. It must implement
+// the mb.MetricSet interface. And this is best achieved by embedding
+// mb.BaseMetricSet because it implements all of the required mb.MetricSet
+// interface methods except for Fetch.
+type MetricSet struct {
+	mb.BaseMetricSet
+	config config
+}
+
+type config struct {
+	Metrics             []string `config:"stackdriver.metrics" validate:"required"`
+	Zone                string   `config:"zone" validate:"required"`
+	ProjectID           string   `config:"project_id" validate:"required"`
+	ExcludeLabels       bool     `config:"exclude_labels"`
+	ServiceName         string   `config:"stackdriver.service"  validate:"required"`
+	CredentialsFilePath string   `config:"credentials_file_path"`
+
+	credentialsOption option.ClientOption
+	opt               []option.ClientOption
+}
+
+// New creates a new instance of the MetricSet. New is responsible for unpacking
+// any MetricSet specific configuration options if there are any.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Beta("The gcp '%s' metricset is beta.", MetricsetName)
+
+	m := &MetricSet{BaseMetricSet: base}
+
+	if err := base.Module().UnpackConfig(&m.config); err != nil {
+		return nil, err
+	}
+
+	m.config.credentialsOption = option.WithCredentialsFile(m.config.CredentialsFilePath)
+
+	if err := validatePeriodForGCP(m.Module().Config().Period); err != nil {
+		return nil, err
+	}
+
+	return m, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) (err error) {
+	m.config.opt = []option.ClientOption{m.config.credentialsOption}
```

Comment:
We could create the final options array in `New`, so we don't keep two variables for options, I don't think this is expected to change on fetch time.

Reply:
Fixed